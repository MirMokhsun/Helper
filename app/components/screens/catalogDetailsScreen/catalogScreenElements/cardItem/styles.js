import { StyleSheet, I18nManager } from 'react-native';

export const styles = StyleSheet.create({
    container: {
        width: '90%',
        height: 75,
        borderRadius: 10,
        backgroundColor: '#ffffff',
        marginBottom: 15,
        alignItems: 'center',
        justifyContent: 'center',
        alignSelf: 'center',
    },
    wrapper: {
        borderLeftWidth: 1,
        height: 40,
        width: '100%',
        flexDirection: 'row',
        alignItems: 'center',
        justifyContent: 'space-around',
    },
    starWrapper: {
        width: 70,
        alignItems: 'center',
        justifyContent: 'center',
    },
    arrowWrapper: {
        marginLeft: 10,
        width: 40,
        alignItems: 'center',
        justifyContent: 'center',
    },
    fullNameText: {
        fontFamily: 'Roboto-Light',
        lineHeight: 32,
        fontSize: 14,
        color: '#000756',
        flex: 2,
        textAlign: I18nManager.isRTL ? 'right' : 'left',
    },
    quotationContainer: {
        flex: 3,
        flexDirection: 'column',
    }
})