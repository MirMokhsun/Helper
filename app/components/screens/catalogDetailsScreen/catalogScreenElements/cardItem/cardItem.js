import React, { Component } from 'react';
import { isEqual } from 'lodash';
import { View, Text, TouchableOpacity } from 'react-native';
import { styles } from './styles';
import UpArrow from '../../../../commonComponents/svg/UpArrow';
import RightArrow from '../../../../commonComponents/svg/RightArrow';
import DownArrow from '../../../../commonComponents/svg/DownArrow';
import { QuotationItem } from '../quotationItem/quotationItem';
import StarSVG from '../../../../commonComponents/svg/starSVG';
import FullStar from '../../../../commonComponents/svg/FullStar';
import { withNavigation } from 'react-navigation';
import { localization } from '../../../../../common/localization';
import { returnObjectLength } from '../../../../../api/funcWithObject/funcWithObject';

export class CardItem extends Component {
    constructor(props) {
        super(props)
        const { pageName, navigation, setQuotationFavorites, cardID, removeFromCatalogDetailScreenCard, getQuotationMap, showToast, titleName } = this.props;
        this.pageName = pageName;
        this.titleName = titleName;
        this.setQuotationFavorites = setQuotationFavorites;
        this.cardID = cardID;
        this.removeFromCatalogDetailScreenCard = removeFromCatalogDetailScreenCard;
        this.getQuotationMap = getQuotationMap;
        this.showToast = showToast;
        this.svgWithBool = { favoriteSvg: <FullStar />, isCardFavorite: true };
        this.choiseId = { key: this.cardID, id: this.cardID };
        this.state = {
            borderColor: '#9FA0B2',
            arrow: <RightArrow />,
            favoriteSvg: <StarSVG />,
            isCardFavorite: false,
            inFocus: true,
            showCard: true
        }
        this.recimendation = {
            'Buy': { borderColor: '#1459D2', arrow: <UpArrow /> },
            'Sell': { borderColor: '#F86F6F', arrow: <DownArrow /> },
            'Neutral': { borderColor: '#9FA0B2', arrow: <RightArrow /> }
        }
        this.addListenerWillFocus = navigation.addListener('willFocus', () => { this.setState({ inFocus: true, }); })
        this.addListenerDidBlur = navigation.addListener('didBlur', () => { this.setState({ inFocus: false, }); })
    }
    
    componentWillMount() {
        const { indicator, isCardFavorite } = this.props;
        this.setColor(indicator);
        if (isCardFavorite) {
            this.setState({ isCardFavorite, favoriteSvg: <FullStar /> });
        }
    }

    componentWillReceiveProps(nextProps) {
        const { indicator } = nextProps;
        this.setColor(indicator);
    }

    shouldComponentUpdate(nextProps, nextState) {
        const { item } = this.props;
        const { isCardFavorite, inFocus } = this.state;
        const isItemChanged = !isEqual(nextProps.item, item);
        const isStateChanged = !isEqual(this.state, nextState);
        const isFavoritesChanged = !isEqual(isCardFavorite, nextState.isCardFavorite)
        return ((isItemChanged || isStateChanged) && inFocus) || isFavoritesChanged;
    }

    componentWillUnmount() {
        this.addListenerWillFocus.remove();
        this.addListenerDidBlur.remove();
    }

    setColor = (indicator) => {
        this.setState(this.recimendation[indicator])
    }

    choseCardFaforite = () => {
        try {
            const { isCardFavorite: isCardFavoriteState } = this.state;
            if (isCardFavoriteState) {
                this.ifCardFavoriteStateTrueFunction();
            } else {
                this.svgWithBool.favoriteSvg = <FullStar />;
                this.svgWithBool.isCardFavorite = true;
                this.choiseId.id = this.cardID; // add
            }
            this.setState(this.svgWithBool);
            this.setQuotationFavorites(this.choiseId);
        } catch (error) {
        }
    }

    ifCardFavoriteStateTrueFunction = () => {
        const quotationMap = this.getQuotationMap();
        const favoritesObjLenght = returnObjectLength(quotationMap.favoritesObj);

        if (favoritesObjLenght > 1) {
            this.svgWithBool.favoriteSvg = <StarSVG />;
            this.svgWithBool.isCardFavorite = false;
            this.choiseId.id = ''; // delete
            this.hideCard();
        } else {
            const { language } = this.props;
            const currenLanguage = localization[language];
            this.showToast(currenLanguage['Last tool cannot be removed from favorites']);
        }
    }

    hideCard = () => {
        if (this.pageName === 'favorites') {
            this.removeFromCatalogDetailScreenCard(this.cardID);
        }
    }

    routeToDetailCardScreen = () => {
        const { navigation, setCardItem, item } = this.props;
        const { isCardFavorite } = this.state;
        setCardItem(item);
        navigation.navigate('DetailCardScreen', { choseCardFaforite: this.choseCardFaforite, isCardFavorite, getQuotationMap: this.getQuotationMap, title: this.titleName });
    }

    render() {
        const { item: { Ask, Bid, FullName } } = this.props;
        const { borderColor, arrow, favoriteSvg, showCard } = this.state;

        return (
            <TouchableOpacity onPress={this.routeToDetailCardScreen} testID={`CardItem${FullName}`} accessibilityLabel={`CardItem${FullName}`}>
                {showCard ? (
                    <View style={styles.container}>
                        <View style={[styles.wrapper, { borderLeftColor: borderColor }]}>
                            <View style={styles.arrowWrapper}>
                                {arrow}
                            </View>
                            <Text style={styles.fullNameText} numberOfLines={1}>{FullName}</Text>
                            <View style={styles.quotationContainer}>
                                <QuotationItem text="Ask" number={(Ask == 0) ? '--' : Ask} />
                                <QuotationItem text="Bid" number={(Bid == 0) ? '--' : Bid} />
                            </View>
                            <TouchableOpacity onPress={this.choseCardFaforite} style={styles.starWrapper} testID={`CardItemFavorit${FullName}`} accessibilityLabel={`CardItemFavorit${FullName}`}>
                                {favoriteSvg}
                            </TouchableOpacity>
                        </View>
                    </View >
                ) : null}
            </TouchableOpacity>
        );
    }
}

export default withNavigation(CardItem);