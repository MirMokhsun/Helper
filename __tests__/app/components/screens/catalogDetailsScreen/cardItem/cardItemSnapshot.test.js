import 'react-native';
import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import { CardItem } from '../../../../../../app/components/screens/catalogDetailsScreen/catalogScreenElements/cardItem/cardItem';

const mock = {
    "ID": 1,
    "baseCurrency": "AUD",
    "defaultQuantity": 10000000000,
    "keyName": "AUDCAD",
    "leverage": 200,
    "Market": "Currency",
    "maximumQuantity": 40000000000000,
    "minimumQuantity": 5000000000,
    "name": "AUD/CAD",
    "popular": false,
    "priceRounding": 100000,
    "quantityIncrement": 5000000000,
    "termCurrency": "CAD",
    "tickSize": 100,
    "Ask": "1.1430",
    "Bid": "0.0000",
    "Diff": "0.0000",
    "ESV": "0.0000",
    "FullName": "AUD/CAD",
    "Max": "0.0000",
    "Min": "0.0000",
    "Per": 0,
    "Rate": "0.0000",
    "SSV": "0.0000",
    "Sess": "---",
    "Timestamp": 0
};

const nav = {
    addListener: () => { }
}

it('create snapshot QuotationItem', () => {
    const renderer = new ShallowRenderer()
    let toolbarSnapshot = renderer.render(<CardItem indicator={'buy'} item={mock} language={'ru'} setCardItem={() => { }} navigation={nav} />);
    expect(toolbarSnapshot).toMatchSnapshot();
});