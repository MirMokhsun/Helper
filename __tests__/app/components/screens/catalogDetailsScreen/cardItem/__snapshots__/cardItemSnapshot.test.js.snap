// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create snapshot QuotationItem 1`] = `
<TouchableOpacity
  accessibilityLabel="CardItemAUD/CAD"
  activeOpacity={0.2}
  onPress={[Function]}
  testID="CardItemAUD/CAD"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "#ffffff",
        "borderRadius": 10,
        "height": 75,
        "justifyContent": "center",
        "marginBottom": 15,
        "width": "90%",
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderLeftWidth": 1,
            "flexDirection": "row",
            "height": 40,
            "justifyContent": "space-around",
            "width": "100%",
          },
          Object {
            "borderLeftColor": "#9FA0B2",
          },
        ]
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "justifyContent": "center",
            "marginLeft": 10,
            "width": 40,
          }
        }
      >
        <RightArrow />
      </View>
      <Text
        numberOfLines={1}
        style={
          Object {
            "color": "#000756",
            "flex": 2,
            "fontFamily": "Roboto-Light",
            "fontSize": 14,
            "lineHeight": 32,
            "textAlign": "left",
          }
        }
      >
        AUD/CAD
      </Text>
      <View
        style={
          Object {
            "flex": 3,
            "flexDirection": "column",
          }
        }
      >
        <QuotationItem
          number="1.1430"
          text="Ask"
        />
        <QuotationItem
          number="--"
          text="Bid"
        />
      </View>
      <TouchableOpacity
        accessibilityLabel="CardItemFavoritAUD/CAD"
        activeOpacity={0.2}
        onPress={[Function]}
        style={
          Object {
            "alignItems": "center",
            "justifyContent": "center",
            "width": 70,
          }
        }
        testID="CardItemFavoritAUD/CAD"
      >
        <StarSVG />
      </TouchableOpacity>
    </View>
  </View>
</TouchableOpacity>
`;
